internal/dag: enforce root ingressroutes live in specific namespaces

alexbrand opened the pull request:
Implements enforcing mode, in which root ingress routes can only be defined in specific namespaces.

If contour encounters a root ingress route that is defined outside of the blessed namespaces, the ingress route is ignored.

If the list of blessed namespaces is empty, root ingress routes can be defined in any namespace.

Fixes #461

alexbrand commented on internal/dag/dag.go:
```diff
@@ -30,6 +30,11 @@ import (
+	// IngressRouteRootNamespaces specifies the namespaces where root
+	// IngressRoutes can be defined. If empty, roots can be defined in any
+	// namespace.
+	IngressRouteRootNamespaces []string
```
I am wondering if there's a better place for this, as this is adding business logic that doesn't have much to do with building a DAG.

alexbrand replied:
> Adding and having to remove later
Not sure I am following.

I had thought about building a ResourceEventHandler that would wrap the DAG and implement this logic. Is that what you have in mind?

stevesloka replied:
What I meant is you could load up the DAG with all the items, then have another process validate that those items should exist in there. I think it's better to remove up front rather than post-processing. 

The same issues will come up later with validating duplicate healthchecks or strategy, etc. Do we look for those while creating the DAG or somehow after.

alexbrand replied:
OK, I think I am with you, and I think that is what I am doing in this case, right? When we are recomputing the DAG, we are checking whether the IngressRoute can exist in the given namespace. In the case that it cannot, we skip processing it and thus do not add it to the DAG.

stevesloka replied:
yup 100% you are doing that! Was just replying to your comment about "Is this the right place for it".

alexbrand replied:
Hehe cool! 

I still think we might want to put something between the watchers and the DAG, and move the validation logic there. Not sure if it's worth it though.

davecheney replied:
This probably should live in the same place that we do a check for ingress class .. but we don't do that at all yet.

I'd try pushing that logic into `contour.DAGAdapter`.

alexbrand commented on cmd/contour/contour.go:
```diff
@@ -98,6 +98,9 @@ func main() {
 	serve.Flag("ingress-class-name", "Contour IngressClass name").StringVar(&t.IngressClass)
 
+	var rootNamespaces []string
+	serve.Flag("ingressroute-root-namespaces", "Restrict contour to searching these namespaces for root ingress routes").StringsVar(&rootNamespaces)
```
Yep, they are. Should probably call that out in the help text.

davecheney commented on cmd/contour/contour.go:
```diff
@@ -123,6 +126,8 @@ func main() {
 
 		// setup DAG watcher and debug handler
 		var dagrh dag.ResourceEventHandler
```
Please move the declaration of `dagrh` above line 101 then you can do 
```
serve.Flag("ingressroute-root-namespaces", "Restrict contour to searching these namespaces for root ingress routes").StringsVar(&dagrh.DAG.IngressRouteRootNamespaces)
```

davecheney commented on internal/dag/dag.go:
```diff
@@ -297,6 +302,11 @@ func (d *DAG) recompute() *dag {
 
+		// ensure root ingressroute lives in allowed namespace
+		if !d.rootAllowed(ir) {
+			continue
```
For status we may have to accept all ingressroutes, then mark the ones that don't match a permitted namespace as orphaned or excluded.

alexbrand replied:
Setting status instead of logging a message would be my preferred approach as well. I am thinking we can tackle that once the Status PR lands.

alexbrand commented:
Addressed review comments and added e2e tests. The RDS tests are failing because RDS is not yet driven by the DAG. We can take them out, or we can merge this once RDS transitions to the DAG.

rosskukulinski commented:
I believe this is currently blocked on #454 so that RDS related tests pass.
